keps/sig-auth/4193-tokenrequest-requester-info-enhancements/README.md

```diff
@@ -0,0 +1,616 @@
+
+The SA authenticator will also be extended to check whether a JWT's embedded node information is still valid/current,
+and if not, will reject the review/indicate to the client that the token mismatches with the current state of Nodes.
+
+For tokens bound to Pod objects, to avoid unexpected breaking changes in behaviour, this will not only be protected by
+the feature flag whilst the feature is graduating to GA, but will also be gated behind an additional kube-apiserver flag
+`--service-account-validate-node-info` which defaults to 'off'. This avoids changing the previous behaviour for
+validation of tokens issued for pods.
+
+Tokens that are directly bound to Node objects will always validate the name and UID, as binding tokens to Node objects
+is a new option and therefore enforcing this validation check from day 1 is non-breaking.
+
+The validation will involve first checking whether the Node object with the name given in the JWT still exists, and if
+it does, validating whether the UID of that Node is equal to the UID embedded in the token.
+
+For pods, this behaviour matches that of the existing `pod-garbage-collector-controller`, which will automatically
+delete Pod  objects that are bound to Node's that no longer exist after a period of time. With this enabled, tokens will
+be immediately invalidated rather than waiting until [the 40s grace period](https://github.com/kubernetes/kubernetes/blob/fc786dcd1d2efcc241e0e2392086934f2806555d/pkg/controller/podgc/gc_controller.go#L50-L52) passes.
+
+This means administrators can **delete node objects to invalidate all JWTs issued that embed that Node's info**.
+
+### Including a UUID ([JTI](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.7)) on each issued JWT
+
+When a TokenRequest is being issued/fulfilled, we will modify the issuing code to also generate and embed a UUID which
+can be later used to trace the requests that a specific issued token has made to the apiserver via the audit log.
+
+This will require changing the JWT issuing code to actually generate this UUID, as well as extending the code around the
+audit log to have it record this information into audit entries.
+
+### User Stories (Optional)
+
+#### Story 1
+
+Alice is building an identity issuance system that relies on verifying projected service account tokens to ensure that
+a pod exists, and is associated with a particular service account.
+
+Alice wants to prevent a malicious actor being able to replay this token from another host, so that she can be certain
+that the user requesting the identity document is also the same user that initially requested the token (as the identity
+used to request the token requires validations rooted in the cryptographic root of trust on the host, e.g. a TPM).
+
+This not only allows us to prevent replay attacks, but allows us to issue identity documents that we can 'root' in the
```

Comment:
This isn't so obvious to me.  It seems like anyone holding a client certificate that looks like the node in question would be able to request JWTs.

Reply:
Yep that is correct - perhaps I should clarify that point... what we are piggy backing on is the process through which a host acquires that certificate. That may be via the CSR API, or it could be some external system that performs its own cross-reference checks. This is the process/attestation that we want to anchor our trust to here :)

Reply:
I think it's worth clarification.  There is no guarantee of a single CSR issuer.  Impersonation could do similar things.  As could front-proxy certificate assertion.  The kube-apiserver authentication stack isn't as restrictive as the TPM module is and there are many ways to assert identity.